Data Loading and Analysis

#### By: Bernales Camila
*****************************

### Step 1 - Loading a Tabular Data File

Make sure the "titanic.csv" file (provided as part of this excercise) is located in the JupyterLab working directory. <br> Read the "titanic.csv" file dataset into a dataframe called df. Print the first 5 rows in df.

In [80]:
import pandas as pd

df = pd.read_csv('/Users/Camila Bernales/Documents/datascience/udemyCourse/archivos/titanic.csv', header= 0)
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket". Print the first 5 rows in df.

In [21]:
df = pd.read_csv('/Users/Camila Bernales/Documents/datascience/udemyCourse/archivos/titanic.csv', header= 0, usecols = [0,1,2,3,4,5,9,10,11])
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age     Fare  \
0                            Braund, Mr. Owen Harris    male  22.0   7.2500   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  71.2833   
2                             Heikkinen, Miss. Laina  female  26.0   7.9250   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  53.1000   
4                           Allen, Mr. William Henry    male  35.0   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

Print the data type of each column by using the dtypes attribute.

In [23]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket" and also convert the column "Survived" into a boolean data type. Print the first 5 rows in df.

In [39]:
df = pd.read_csv('/Users/Camila Bernales/Documents/datascience/udemyCourse/archivos/titanic.csv', header= 0, usecols=[0,1,2,3,4,5,9,10,11], dtype= {'Survived': 'Bool'})
df.dtypes

PassengerId      int64
Survived          bool
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

Print the data type of each column by using the dtypes attribute. The "Survived" should be boolean data type.

In [40]:
df.dtypes

PassengerId      int64
Survived          bool
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

### Step 2 - Preview the DataFrame

Print the two index components of the df dataframe: index, columns.

In [42]:
print(df.index)
print(df.columns)

RangeIndex(start=0, stop=891, step=1)
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'Fare',
       'Cabin', 'Embarked'],
      dtype='object')


Print the size of the dataframe using the shape attribute. We have 891 rows and 9 columns. 

In [43]:
df.shape

(891, 9)

Print the amount of rows per each column without missing values.

In [44]:
df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
Fare           891
Cabin          204
Embarked       889
dtype: int64

Print the unique values for the "Sex" column.

In [48]:
df['Sex'].unique()

array(['male', 'female'], dtype=object)

Print the unique values for the "Embarked" column.

In [49]:
df['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

Print the amount of unique values for the "Embarked" column.

In [50]:
len(df['Embarked'].unique())

4

Print the number of rows with the same specific value in the "Embarked" column.

In [55]:
df['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

How many passengers survived?

In [57]:
df['Survived'].value_counts()

False    549
True     342
Name: Survived, dtype: int64

What is the survival percentage?

In [60]:
df['Survived'].value_counts(normalize =  True)

False    0.616162
True     0.383838
Name: Survived, dtype: float64

### Step 3 - Using Summary Statistics

What is the maximum age of the titanic passengers?

In [61]:
df['Age'].max()

80.0

What is the average age and average ticket price?

In [62]:
df[["Age", "Fare"]].mean()

Age     29.699118
Fare    32.204208
dtype: float64

How many passengers survived? (using the sum method).

In [65]:
df['Survived'].sum()

342

Print a full statistical summary using the describe method. What is the agev value that 25% of passengers are smaller then this age? (should be 20.123).

In [66]:
df.describe()

PassengerId      Pclass         Age        Fare
count   891.000000  891.000000  714.000000  891.000000
mean    446.000000    2.308642   29.699118   32.204208
std     257.353842    0.836071   14.526497   49.693429
min       1.000000    1.000000    0.420000    0.000000
25%     223.500000    2.000000   20.125000    7.910400
50%     446.000000    3.000000   28.000000   14.454200
75%     668.500000    3.000000   38.000000   31.000000
max     891.000000    3.000000   80.000000  512.329200

### Step 4 - Sorting and Ranking

Sort all rows in the dataframe using the 'Age' column is descending order and print the first 5 rows. 

In [70]:
df.sort_values('Age', ascending=False).head(5)

PassengerId  Survived  Pclass                                  Name  \
630          631      True       1  Barkworth, Mr. Algernon Henry Wilson   
851          852     False       3                   Svensson, Mr. Johan   
493          494     False       1               Artagaveytia, Mr. Ramon   
96            97     False       1             Goldschmidt, Mr. George B   
116          117     False       3                  Connors, Mr. Patrick   

      Sex   Age     Fare Cabin Embarked  
630  male  80.0  30.0000   A23        S  
851  male  74.0   7.7750   NaN        S  
493  male  71.0  49.5042   NaN        C  
96   male  71.0  34.6542    A5        C  
116  male  70.5   7.7500   NaN        Q

Sort all rows in the dataframe using the 'Age' column is descending order and save it in the same dataframe. Print the first 5 rows. 

In [77]:
df.sort_values('Age', ascending = False, inplace = True)
df.head(5)

PassengerId  Survived  Pclass                                  Name  \
630          631      True       1  Barkworth, Mr. Algernon Henry Wilson   
851          852     False       3                   Svensson, Mr. Johan   
493          494     False       1               Artagaveytia, Mr. Ramon   
96            97     False       1             Goldschmidt, Mr. George B   
116          117     False       3                  Connors, Mr. Patrick   

      Sex   Age     Fare Cabin Embarked  
630  male  80.0  30.0000   A23        S  
851  male  74.0   7.7750   NaN        S  
493  male  71.0  49.5042   NaN        C  
96   male  71.0  34.6542    A5        C  
116  male  70.5   7.7500   NaN        Q

Sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [83]:
df.sort_values(['Sex', 'Age'], ascending=[True, True])[['Sex', 'Name', 'Age']]

Sex                               Name   Age
469  female      Baclini, Miss. Helene Barbara  0.75
644  female             Baclini, Miss. Eugenie  0.75
172  female       Johnson, Miss. Eleanor Ileen  1.00
381  female        Nakid, Miss. Maria ("Mary")  1.00
119  female  Andersson, Miss. Ellis Anna Maria  2.00
..      ...                                ...   ...
839    male               Marechal, Mr. Pierre   NaN
846    male           Sage, Mr. Douglas Bullen   NaN
859    male                   Razi, Mr. Raihed   NaN
868    male        van Melkebeke, Mr. Philemon   NaN
878    male                 Laleff, Mr. Kristo   NaN

[891 rows x 3 columns]

Print the row in location 2 using the iloc method.

In [85]:
df.iloc[[2]]

PassengerId  Survived  Pclass                    Name     Sex   Age  SibSp  \
2            3         1       3  Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket   Fare Cabin Embarked  
2      0  STON/O2. 3101282  7.925   NaN        S

Drop the missing values and then sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [84]:
df.dropna().sort_values(['Sex', 'Age'], ascending=[True, True])[[ 'Sex', 'Name', 'Age']]

Sex                                  Name   Age
205  female            Strom, Miss. Telma Matilda   2.0
297  female          Allison, Miss. Helen Loraine   2.0
10   female       Sandstrom, Miss. Marguerite Rut   4.0
618  female           Becker, Miss. Marion Louise   4.0
435  female             Carter, Miss. Lucile Polk  14.0
..      ...                                   ...   ...
54     male        Ostby, Mr. Engelhart Cornelius  65.0
456    male             Millet, Mr. Francis Davis  65.0
745    male          Crosby, Capt. Edward Gifford  70.0
96     male             Goldschmidt, Mr. George B  71.0
630    male  Barkworth, Mr. Algernon Henry Wilson  80.0

[183 rows x 3 columns]

What are the five oldest passengers that paid a ticket that cost more than the average ticket price? (create a boolean condition and than use it to filter the dataframe and than use the nlargest.

In [94]:
filter = df['Fare']>df['Fare'].mean()
df[filter].nlargest(5, 'Age')[['Name', 'Age', 'Sex']]

Name   Age   Sex
96        Goldschmidt, Mr. George B  71.0  male
493         Artagaveytia, Mr. Ramon  71.0  male
745    Crosby, Capt. Edward Gifford  70.0  male
54   Ostby, Mr. Engelhart Cornelius  65.0  male
438               Fortune, Mr. Mark  64.0  male

### Step 5 - Filtering

Filter and keep only the male passengers.

In [97]:
df[df['Sex'] == 'male']

PassengerId  Survived  Pclass                            Name   Sex  \
0              1         0       3         Braund, Mr. Owen Harris  male   
4              5         0       3        Allen, Mr. William Henry  male   
5              6         0       3                Moran, Mr. James  male   
6              7         0       1         McCarthy, Mr. Timothy J  male   
7              8         0       3  Palsson, Master. Gosta Leonard  male   
..           ...       ...     ...                             ...   ...   
883          884         0       2   Banfield, Mr. Frederick James  male   
884          885         0       3          Sutehall, Mr. Henry Jr  male   
886          887         0       2           Montvila, Rev. Juozas  male   
889          890         1       1           Behr, Mr. Karl Howell  male   
890          891         0       3             Dooley, Mr. Patrick  male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
4    35.0      0      0            373450   8.0500   NaN        S  
5     NaN      0      0            330877   8.4583   NaN        Q  
6    54.0      0      0             17463  51.8625   E46        S  
7     2.0      3      1            349909  21.0750   NaN        S  
..    ...    ...    ...               ...      ...   ...      ...  
883  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
886  27.0      0      0            211536  13.0000   NaN        S  
889  26.0      0      0            111369  30.0000  C148        C  
890  32.0      0      0            370376   7.7500   NaN        Q  

[577 rows x 12 columns]

Filer out all passengers below the age of 40 and that are also 'male'. Print the ten first results.

In [102]:
df[(df['Age']<40) & (df['Sex'] == 'male')].head(10)

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0             1         0       3         Braund, Mr. Owen Harris  male  22.0   
4             5         0       3        Allen, Mr. William Henry  male  35.0   
7             8         0       3  Palsson, Master. Gosta Leonard  male   2.0   
12           13         0       3  Saundercock, Mr. William Henry  male  20.0   
13           14         0       3     Andersson, Mr. Anders Johan  male  39.0   
16           17         0       3            Rice, Master. Eugene  male   2.0   
20           21         0       2            Fynney, Mr. Joseph J  male  35.0   
21           22         1       2           Beesley, Mr. Lawrence  male  34.0   
23           24         1       1    Sloper, Mr. William Thompson  male  28.0   
27           28         0       1  Fortune, Mr. Charles Alexander  male  19.0   

    SibSp  Parch     Ticket     Fare        Cabin Embarked  
0       1      0  A/5 21171    7.250          NaN        S  
4       0      0     373450    8.050          NaN        S  
7       3      1     349909   21.075          NaN        S  
12      0      0  A/5. 2151    8.050          NaN        S  
13      1      5     347082   31.275          NaN        S  
16      4      1     382652   29.125          NaN        Q  
20      0      0     239865   26.000          NaN        S  
21      0      0     248698   13.000          D56        S  
23      0      0     113788   35.500           A6        S  
27      3      2      19950  263.000  C23 C25 C27        S

### Step 6 - Grouping and Aggregating

What is the average ticket price for male versus female passengers?

In [105]:
df[['Sex', 'Fare']].groupby('Sex').mean()

Fare
Sex              
female  44.479818
male    25.523893

What is the average age for the survived group?

In [114]:
df.groupby('Survived').mean()

PassengerId    Pclass        Age     SibSp     Parch       Fare
Survived                                                                 
0          447.016393  2.531876  30.626179  0.553734  0.329690  22.117887
1          444.368421  1.950292  28.343690  0.473684  0.464912  48.395408

How many survived group by the embarked port?

In [122]:
df.groupby('Embarked')['Survived'].mean()

Embarked
C    0.553571
Q    0.389610
S    0.336957
Name: Survived, dtype: float64

What is the average age as well as the average price for male and female?

In [124]:
df.groupby('Sex').mean()

PassengerId  Survived    Pclass        Age     SibSp     Parch  \
Sex                                                                      
female   431.028662  0.742038  2.159236  27.915709  0.694268  0.649682   
male     454.147314  0.188908  2.389948  30.726645  0.429809  0.235702   

             Fare  
Sex                
female  44.479818  
male    25.523893

 count, mean, max and min of number of passengers boarding by age

In [126]:
df.groupby('Embarked')['Age'].agg(['count', 'mean', 'max', 'min'])

count       mean   max   min
Embarked                              
C           130  30.814769  71.0  0.42
Q            28  28.089286  70.5  2.00
S           554  29.445397  80.0  0.67

### End of document
************************